### PR TITLE
ci(debugging): add exploration testing

### PR DESCRIPTION
**Pull request by P403n1x87**
## Description

This change adds the so-called exploration tests which we use to ensure that the dynamic instrumentation does not affect the expected behaviour of the instrumented code. We select a few popular Python frameworks and inject probes on each single line and wrap every single function. We then run the full test suite for the actual validation.

At the end of the test run, line coverage and function call counts are reported as a "proof-of-work"

~~~
===================== DeterministicProfiler: probes stats ======================

Installed probes: 220/220

============================== Function coverage ===============================

Functions called: 174/232

Function                                                             Calls
==========================================================================
requests.structures.CaseInsensitiveDict.__setitem__                   8128
requests.structures.CaseInsensitiveDict.__getitem__                   7172
requests.structures.CaseInsensitiveDict.__init__                      2129
requests.cookies.MockResponse.info                                    1576
requests.utils.to_key_val_list                                        1533
requests.structures.CaseInsensitiveDict.__iter__                      1530
requests._internal_utils.to_native_string                             1462
requests.sessions.merge_setting                                       1359
requests.cookies.MockRequest.__init__                                 1243
requests.cookies.MockRequest.get_full_url                             1178
requests.utils.should_bypass_proxies                                  1092
requests.hooks.default_hooks                                          1023
requests.utils.check_header_validity                                   846
requests.cookies.cookiejar_from_dict                                   835
requests.cookies.extract_cookies_to_jar                                793
requests.cookies.MockResponse.__init__                                 788
requests.structures.CaseInsensitiveDict.copy                           765
requests.cookies.RequestsCookieJar.update                              727
requests.structures.CaseInsensitiveDict.__len__                        639
requests.utils.get_environ_proxies                                     637
requests.utils.select_proxy                                            623
requests.cookies.merge_cookies                                         559
requests.adapters.HTTPAdapter.init_poolmanager                         503
requests.adapters.HTTPAdapter.__init__                                 501
requests.adapters.BaseAdapter.__init__                                 501
requests.sessions.Session.mount                                        498
requests.models.PreparedRequest.__init__                               478
requests.models.PreparedRequest.prepare_cookies                        455
requests.cookies.get_cookie_header                                     455
requests.cookies.MockRequest.get_new_headers                           455
requests.models.Response.content.getter                                439
requests.utils.unquote_unreserved                                      429
requests.sessions.SessionRedirectMixin.rebuild_proxies                 429
requests.utils.requote_uri                                             427
requests.cookies.MockRequest.unverifiable.getter                       387
requests.cookies.MockRequest.origin_req_host.getter                    387
requests.cookies.MockRequest.is_unverifiable                           387
requests.cookies.MockRequest.get_origin_req_host                       387
requests.cookies.MockRequest.get_host                                  387
requests.models.Response.is_redirect.getter                            379
requests.cookies.RequestsCookieJar.set_cookie                          358
requests.models.RequestHooksMixin.register_hook                        342
requests.sessions.SessionRedirectMixin.get_redirect_target             329
requests.adapters.HTTPAdapter.close                                    326
requests.models.PreparedRequest.prepare_content_length                 320
requests.utils.get_netrc_auth                                          308
requests.adapters.HTTPAdapter.send                                     308
requests.adapters.HTTPAdapter.get_connection                           308
requests.models.RequestEncodingMixin._encode_params                    307
requests.models.PreparedRequest.prepare_url                            305
requests.models.PreparedRequest.prepare_method                         305
requests.models.PreparedRequest.prepare                                305
requests.adapters.HTTPAdapter.cert_verify                              303
requests.models.RequestEncodingMixin.path_url.getter                   301
requests.adapters.HTTPAdapter.request_url                              300
requests.adapters.HTTPAdapter.add_headers                              300
requests.models.PreparedRequest.prepare_headers                        297
requests.models.Request.__init__                                       296
requests._internal_utils.unicode_is_ascii                              293
requests.sessions.Session.get_adapter                                  290
requests.models.PreparedRequest.prepare_body                           289
requests.models.Response.__init__                                      287
requests.sessions.Session.send                                         286
requests.models.PreparedRequest.prepare_hooks                          284
requests.models.PreparedRequest.prepare_auth                           284
requests.utils.get_auth_from_url                                       278
requests.models.Response.iter_content                                  276
requests.utils.get_encoding_from_headers                               274
requests.adapters.HTTPAdapter.build_response                           271
requests.utils._parse_content_type_header                              251
requests.sessions.SessionRedirectMixin.resolve_redirects               249
requests.hooks.dispatch_hook                                           246
requests.utils.default_user_agent                                      242
requests.utils.default_headers                                         242
requests.sessions.Session.__init__                                     242
requests.sessions.merge_hooks                                          210
requests.sessions.Session.prepare_request                              210
requests.sessions.Session.request                                      198
requests.sessions.Session.merge_environment_settings                   182
requests.models.PreparedRequest.copy                                   172
requests.cookies._copy_cookie_jar                                      172
requests.cookies.RequestsCookieJar.get_policy                          172
requests.cookies.RequestsCookieJar.copy                                171
requests.models.Response.close                                         169
requests.sessions.Session.close                                        163
requests.sessions.Session.__exit__                                     162
requests.sessions.Session.__enter__                                    162
requests.api.request                                                   162
requests.sessions.SessionRedirectMixin.rebuild_method                  139
requests.sessions.SessionRedirectMixin.rebuild_auth                    139
requests.api.get                                                       124
requests.utils.unquote_header_value                                    113
requests.utils.is_valid_cidr                                           100
requests.models.Request.prepare                                         84
requests.utils.super_len                                                71
requests.exceptions.RequestException.__init__                           67
requests.utils.is_ipv4_address                                          52
requests.cookies.MockRequest.has_header                                 41
requests.cookies.MockRequest.add_unredirected_header                    41
requests.sessions.Session.get                                           34
requests.auth.HTTPDigestAuth.build_digest_header                        33
requests.structures.CaseInsensitiveDict.lower_items                     32
requests.cookies.create_cookie                                          31
requests.auth.HTTPDigestAuth.handle_redirect                            31
requests.auth.HTTPDigestAuth.handle_401                                 31
requests.auth.HTTPDigestAuth.init_per_thread_state                      30
requests.auth.HTTPDigestAuth.__call__                                   30
requests.api.post                                                       30
requests.utils.parse_dict_header                                        29
requests.utils.dotted_netmask                                           29
requests.utils.address_in_network                                       26
requests.auth._basic_auth_str                                           24
requests.auth.HTTPDigestAuth.__init__                                   24
requests.sessions.session                                               21
requests.models.RequestEncodingMixin._encode_files                      21
requests.cookies.RequestsCookieJar._find_no_duplicates                  21
requests.utils.prepend_scheme_if_needed                                 20
requests.structures.CaseInsensitiveDict.__delitem__                     19
requests.models.Response.text.getter                                    18
requests.structures.CaseInsensitiveDict.__eq__                          17
requests.cookies.RequestsCookieJar.__getitem__                          17
requests.cookies.RequestsCookieJar.set                                  16
requests.auth.HTTPBasicAuth.__init__                                    16
requests.auth.HTTPBasicAuth.__call__                                    16
requests.utils.guess_filename                                           15
requests.sessions.SessionRedirectMixin.should_strip_auth                14
requests.adapters.HTTPAdapter.proxy_manager_for                         14
requests.utils.guess_json_utf                                           13
requests.models.Response.json                                           13
requests.utils.urldefragauth                                            12
requests.utils.extract_zipped_paths                                     12
requests.cookies.RequestsCookieJar.iteritems                            12
requests.cookies.RequestsCookieJar.items                                10
requests.utils.iter_slices                                               8
requests.models.Response.raise_for_status                                8
requests.adapters.HTTPAdapter.proxy_headers                              8
requests.cookies.RequestsCookieJar.__setstate__                          7
requests.cookies.RequestsCookieJar.__getstate__                          7
requests.utils.rewind_body                                               6
requests.utils.parse_header_links                                        6
requests.utils.get_encodings_from_content                                6
requests.structures.LookupDict.__init__                                  6
requests.cookies.morsel_to_cookie                                        6
requests.models.Response.ok.getter                                       4
requests.cookies.RequestsCookieJar.get                                   4
requests.api.head                                                        4
requests.models.Response.iter_lines                                      3
requests.help.info                                                       3
requests.help._implementation                                            3
requests.cookies.RequestsCookieJar.keys                                  3
requests.cookies.RequestsCookieJar.iterkeys                              3
requests.cookies.RequestsCookieJar.__contains__                          3
requests.utils.stream_decode_response_unicode                            2
requests.utils.add_dict_to_cookiejar                                     2
requests.structures.LookupDict.get                                       2
requests.structures.LookupDict.__getitem__                               2
requests.models.Response.__repr__                                        2
requests.api.put                                                         2
requests.adapters.HTTPAdapter.__setstate__                               2
requests.adapters.HTTPAdapter.__getstate__                               2
requests.structures.LookupDict.__repr__                                  1
requests.structures.CaseInsensitiveDict.__repr__                         1
requests.sessions.Session.__setstate__                                   1
requests.sessions.Session.__getstate__                                   1
requests.models.Response.links.getter                                    1
requests.models.Response.apparent_encoding.getter                        1
requests.models.Response.__setstate__                                    1
requests.models.Response.__iter__                                        1
requests.models.Response.__getstate__                                    1
requests.models.Response.__exit__                                        1
requests.models.Response.__enter__                                       1
requests.cookies.RequestsCookieJar.values                                1
requests.cookies.RequestsCookieJar.itervalues                            1
requests.cookies.RequestsCookieJar.__setitem__                           1
requests.utils.parse_list_header                                         0
requests.utils.get_unicode_from_response                                 0
requests.utils.from_key_val_list                                         0
requests.utils.dict_to_sequence                                          0
requests.utils.dict_from_cookiejar                                       0
requests.status_codes._init                                              0
requests.sessions.Session.put                                            0
requests.sessions.Session.post                                           0
requests.sessions.Session.patch                                          0
requests.sessions.Session.options                                        0
requests.sessions.Session.head                                           0
requests.sessions.Session.delete                                         0
requests.models.Response.text.text                                       0
requests.models.Response.ok.ok                                           0
requests.models.Response.next.next                                       0
requests.models.Response.next.getter                                     0
requests.models.Response.links.links                                     0
requests.models.Response.is_redirect.is_redirect                         0
requests.models.Response.is_permanent_redirect.is_permanent_redirect     0
requests.models.Response.is_permanent_redirect.getter                    0
requests.models.Response.content.content                                 0
requests.models.Response.apparent_encoding.apparent_encoding             0
requests.models.Response.__nonzero__                                     0
requests.models.Response.__bool__                                        0
requests.models.RequestHooksMixin.deregister_hook                        0
requests.models.RequestEncodingMixin.path_url.path_url                   0
requests.models.Request.__repr__                                         0
requests.models.PreparedRequest._get_idna_encoded_host                   0
requests.models.PreparedRequest.__repr__                                 0
requests.help.main                                                       0
requests.cookies.remove_cookie_by_name                                   0
requests.cookies.RequestsCookieJar.multiple_domains                      0
requests.cookies.RequestsCookieJar.list_paths                            0
requests.cookies.RequestsCookieJar.list_domains                          0
requests.cookies.RequestsCookieJar.get_dict                              0
requests.cookies.RequestsCookieJar._find                                 0
requests.cookies.RequestsCookieJar.__delitem__                           0
requests.cookies.MockResponse.getheaders                                 0
requests.cookies.MockRequest.unverifiable.unverifiable                   0
requests.cookies.MockRequest.origin_req_host.origin_req_host             0
requests.cookies.MockRequest.host.host                                   0
requests.cookies.MockRequest.host.getter                                 0
requests.cookies.MockRequest.get_type                                    0
requests.cookies.MockRequest.get_header                                  0
requests.cookies.MockRequest.add_header                                  0
requests.check_compatibility                                             0
requests.auth.HTTPProxyAuth.__call__                                     0
requests.auth.HTTPDigestAuth.__ne__                                      0
requests.auth.HTTPDigestAuth.__eq__                                      0
requests.auth.HTTPBasicAuth.__ne__                                       0
requests.auth.HTTPBasicAuth.__eq__                                       0
requests.auth.AuthBase.__call__                                          0
requests.api.patch                                                       0
requests.api.options                                                     0
requests.api.delete                                                      0
requests.adapters.BaseAdapter.send                                       0
requests.adapters.BaseAdapter.close                                      0
requests._check_cryptography                                             0


========================== LineCoverage: probes stats ==========================

Installed probes: 1721/1721

================================ Line coverage =================================

Source                                   Lines Covered
======================================================
requests/__init__.py                        25      0%
requests/__version__.py                      0      0%
requests/_internal_utils.py                 12     92%
requests/adapters.py                       205     94%
requests/api.py                             10     70%
requests/auth.py                           143     85%
requests/certs.py                            0      0%
requests/compat.py                           0      0%
requests/cookies.py                        208     74%
requests/exceptions.py                       7    100%
requests/help.py                            61     66%
requests/hooks.py                           11    100%
requests/models.py                         379     86%
requests/packages.py                         0      0%
requests/sessions.py                       267     96%
requests/status_codes.py                     9      0%
requests/structures.py                      22    100%
requests/utils.py                          362     84%
------------------------------------------------------
TOTAL                                     1721     84%
~~~

## Checklist
- [x] Title must conform to [conventional commit](https://github.com/conventional-changelog/commitlint/tree/master/%40commitlint/config-conventional).
- [x] Add additional sections for `feat` and `fix` pull requests.
- [x] Ensure tests are passing for affected code.
- [x] [Library documentation](https://github.com/DataDog/dd-trace-py/tree/1.x/docs) and/or [Datadog's documentation site](https://github.com/DataDog/documentation/) is updated. Link to doc PR in description.

## Reviewer Checklist
- [ ] Title is accurate.
- [ ] Description motivates each change.
- [ ] No unnecessary changes were introduced in this PR.
- [ ] PR cannot be broken up into smaller PRs.
- [ ] Avoid breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [ ] Tests provided or description of manual testing performed is included in the code or PR.
- [x] Release note has been added for fixes and features, or else `changelog/no-changelog` label added.
- [ ] All relevant GitHub issues are correctly linked.
- [ ] Backports are identified and tagged with Mergifyio.
- [ ] Add to milestone.

### CODE REVIEW

**Comment by P403n1x87** [comment]
#3996 was cherry-picked from here

File: tests/debugging/exploration/debugger.py [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+class ModuleCollector(DebuggerModuleWatchdog):
+    def __init__(self, *args, **kwargs):
+        super(ModuleCollector, self).__init__(*args, **kwargs)
+
+    def on_collect(self, discovery):
+        raise NotImplementedError()
+
+    def after_import(self, module):
+        o = origin(module)
+        if o.startswith(CWD) and not o.startswith(TESTS) and (VENV is None or not o.startswith(VENV)):
```
Comment (majorgreys) [inline tests/debugging/exploration/debugger.py:52]: I wasn't sure where in the code the specific module we'd be interested in is instrumented for the debugger. A comment here would be helpful.

**Review by majorgreys — APPROVED** [review]
lgtm -- the code is a fairly involved use of the debugger which requires one to be familiar with the private interface. Though I raise one comment suggestion, I do wonder how and where to explain the test results for a contributor. Thinking of avoiding noise for contributors as well as unnecessary CI work, should these actions be skipped if the changes do not impact the debugger code or tests?

File: .github/workflows/debugger_exploration_testing.yml [hunk 2]
```diff
@@ -0,0 +1,362 @@
+name: Debugger Exploration Testing
+
+on:
+  push:
+    branches:
+      - master
```
Comment (Kyle-Verhoog) [inline .github/workflows/debugger_exploration_testing.yml:6]:
```suggestion
      - 1.x
```
?

File: .github/workflows/debugger_exploration_testing.yml [hunk 3]
```diff
@@ -0,0 +1,362 @@
...
+          path: ddtrace
+      - uses: actions/checkout@v2
+        with:
+          repository: bottlepy/bottle
+          ref: master
+          path: bottle
+      - name: Install Dependencies
+        run: sudo apt-get install -y libev-dev
+      - name: Test dependencies (Server back-ends and template engines)
+      # Taken from install script inside of .github/workflows of test suite (https://github.com/bottlepy/bottle/blob/master/.github/workflows/run_tests.yml)
```
Comment (Kyle-Verhoog) [inline .github/workflows/debugger_exploration_testing.yml:37]: nit: probably worth getting a perma (commit-hash) link here

File: .github/workflows/debugger_exploration_testing.yml [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+      - name: Install test dependencies
+        run: pip install -e .[test]
+      - name: MarkupSafe fix
+        run: pip install --upgrade MarkupSafe==0.18 pip setuptools --force
+      - name: Disable failing tests
+        run: |
+          sed -i'' "s/test_detect_lang/detect_lang/g" tests/test_units/test_basic_app.py
+          sed -i'' "s/test_langs/langs/g" tests/test_units/test_basic_app.py
+      - name: Run tests
+        run: nosetests
```
Comment (Kyle-Verhoog) [inline .github/workflows/debugger_exploration_testing.yml:362]: There's a lot of pulling the latest branch/release for each of the projects and their dependencies. This seems like it'll be a decent amount of work to maintain.

Do we wanna consider locking all these in?

File: tests/debugging/exploration/_coverage.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
+from collections import defaultdict
```
Comment (Kyle-Verhoog) [inline tests/debugging/exploration/_coverage.py:1]: Worth having type hints?

File: tests/debugging/exploration/__init__.py [hunk 6]
```diff
@@ -0,0 +1,2 @@
+# To run an exploration test, set
+# PYTHONPATH=/path/to/tests/debugging/exploration/
```
Comment (Kyle-Verhoog) [inline tests/debugging/exploration/__init__.py:2]: is it worth configuring this as a runpy module? python -m tests/debugging/exploration?

A `__main__` module could be a little more user friendly than needing to set pythonpath and `DD_DEBUGGER_EXPL_COVERAGE_ENABLED`

File: .github/workflows/debugger_exploration_testing.yml (hunk 3 above)
Comment (P403n1x87) [reply .github/workflows/debugger_exploration_testing.yml:37]: This source was adapted from a verbatim copy of `test_frameworks.yml`. Perhaps we can address these issues in a clean-up PR? Or perhaps consolidate the tests in a single run by enabling exploration tests in `test_frameworks.yml`?

File: tests/debugging/exploration/__init__.py (hunk 6 above)
Comment (P403n1x87) [reply tests/debugging/exploration/__init__.py:2]: All good points, but considering that this is mostly for testing and internal use, I'd be a bit reluctant to add extra complexity here

File: .github/workflows/debugger_exploration_testing.yml (hunk 3 above)
Comment (Kyle-Verhoog) [reply .github/workflows/debugger_exploration_testing.yml:38]: > Or perhaps consolidate the tests in a single run by enabling exploration tests in test_frameworks.yml?

Let's explore this option briefly, I'm ok to not do this and merge this as-is if it is too much overhead to merge the suites

File: .github/workflows/test_frameworks.yml [hunk 7]
```diff
@@ -230,15 +238,18 @@ jobs:
         env:
           # Disabled distributed tracing since there are a lot of tests that assert on headers
           DD_HTTPX_DISTRIBUTED_TRACING: "false"
+          # Debugger exploration testing does not work in CI
+          # PYTHONPATH: ../ddtrace/tests/debugging/exploration/
         # test_pool_timeout raises RuntimeError: The connection pool was closed while 1 HTTP requests/responses were still in-flight
-        run: pytest -k 'not test_pool_timeout'
+        run: pytest -k 'not test_pool_timeout and not test_deleting_unclosed_async_client_causes_warning'
```
Comment (Kyle-Verhoog) [inline .github/workflows/test_frameworks.yml:244]: is this change required still if we're not doing the exploration tests for this job

**Review by Kyle-Verhoog — DISMISSED** [review]

File: .github/workflows/test_frameworks.yml (hunk 7 above)
Comment (P403n1x87) [reply .github/workflows/test_frameworks.yml:244]: ah right! I guess this can be reverted.

**Review by Kyle-Verhoog — APPROVED** [review]

**Review by majorgreys — APPROVED** [review]
